ood-portal-generator/lib/ood_portal_generator/dex.rb

```diff
@@ -0,0 +1,132 @@
+require 'active_support'
+require 'active_support/core_ext'
+require 'digest/sha1'
+require 'fileutils'
+
+module OodPortalGenerator
+  # A view class that renders a Dex configuration
+  class Dex
+    # @param opts [#to_h] the options describing the context used to render the Dex config
+    def initialize(opts = {}, view)
```

Comment:
This method is large and a little hard to follow. One approach to refactoring could be as follows.

```
def initalize(opts = {}, view)
  opts = opts.to_h.each_with_object({}) { |(k, v), h| h[k.to_sym] = v unless v.nil? }
  @config = opts.fetch(:dex, {})
  if @config.nil? || @config == false
    @enable = false
  else
    @config = config.to_h.each_with_object({}) { |(k, v), h| h[k.to_sym] = v unless v.nil? }
    @view = view
  end
end

def ssl?
  @config.fetch(:ssl, !view.ssl.nil?)
end

def protocol
  ssl? ? "https://" : "http://"
end

def servername
  @view.servername || OodPortalGenerator.fqdn
end

def tls_cert
  @config.fetch(:tls_cert, nil)
end

def tls_key
  @config.fetch(:tls_key, nil)
end

def copy_ssl_certs_if_needed
  if ssl? && ! @view.ssl.nil? && tls_cert.nil? && tls_key.nil?
    ...
   end
end

# etc
```

I don't know. Could postpone and open an issue. I see how if Dex is installed, you want to have the View instance to have the OIDC attributes properly modified for Dex and that complicates things a bit. For that I'd probably have the Dex object return a hash of oidc attribute as key value pairs and then do something in generate like view.update_oidc_attributes(dex.oidc_attributes) to decouple the two slightly.

Reply:
I went ahead and made those changes, wasn't too difficult.  I kept the logic for setting `@dex_config` inside initialize which is what the view initializer does too.